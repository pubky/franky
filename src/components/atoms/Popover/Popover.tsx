'use client';

import * as PopoverPrimitive from '@radix-ui/react-popover';
import { useState, useRef, useEffect, createContext, useContext } from 'react';

import * as Libs from '@/libs';
import * as Hooks from '@/hooks';

interface PopoverContextType {
  hover?: boolean;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
}

const PopoverContext = createContext<PopoverContextType>({});

interface PopoverProps extends React.ComponentProps<typeof PopoverPrimitive.Root> {
  hover?: boolean;
  hoverDelay?: number;
}

function Popover({ open, onOpenChange, hover = false, hoverDelay = 0, children, ...props }: PopoverProps) {
  const [internalOpen, setInternalOpen] = useState(false);
  const timeoutRef = useRef<NodeJS.Timeout | null>(null);
  const isTouchDevice = Hooks.useIsTouchDevice();
  const isControlled = open !== undefined;
  const isOpen = isControlled ? open : internalOpen;
  const handleOpenChange = isControlled ? onOpenChange : setInternalOpen;

  // Disable hover on touch devices
  const effectiveHover = hover && !isTouchDevice;

  const handleMouseEnter = () => {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
      timeoutRef.current = null;
    }
    if (effectiveHover) {
      if (hoverDelay > 0) {
        timeoutRef.current = setTimeout(() => {
          handleOpenChange?.(true);
        }, hoverDelay);
      } else {
        handleOpenChange?.(true);
      }
    }
  };

  const handleMouseLeave = () => {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
      timeoutRef.current = null;
    }
    if (effectiveHover) {
      handleOpenChange?.(false);
      // Remove focus from trigger when closing via hover
      setTimeout(() => {
        const activeElement = document.activeElement as HTMLElement;
        if (activeElement && activeElement.blur) {
          activeElement.blur();
        }
      }, 0);
    }
  };

  useEffect(() => {
    return () => {
      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current);
      }
    };
  }, []);

  return (
    <PopoverContext.Provider
      value={{
        hover: effectiveHover,
        onMouseEnter: effectiveHover ? handleMouseEnter : undefined,
        onMouseLeave: effectiveHover ? handleMouseLeave : undefined,
      }}
    >
      <PopoverPrimitive.Root
        data-slot="popover"
        data-testid="popover"
        open={isOpen}
        onOpenChange={handleOpenChange}
        {...props}
      >
        {children}
      </PopoverPrimitive.Root>
    </PopoverContext.Provider>
  );
}

function PopoverTrigger({
  asChild,
  onMouseEnter: propOnMouseEnter,
  onMouseLeave: propOnMouseLeave,
  ...props
}: React.ComponentProps<typeof PopoverPrimitive.Trigger>) {
  const { onMouseEnter: contextOnMouseEnter, onMouseLeave: contextOnMouseLeave, hover } = useContext(PopoverContext);

  const handleMouseEnter = (e: React.MouseEvent<HTMLButtonElement>) => {
    contextOnMouseEnter?.();
    propOnMouseEnter?.(e);
  };

  const handleMouseLeave = (e: React.MouseEvent<HTMLButtonElement>) => {
    contextOnMouseLeave?.();
    propOnMouseLeave?.(e);
  };

  const handleFocus = (e: React.FocusEvent<HTMLButtonElement>) => {
    // Prevent focus outline when using hover mode
    if (hover) {
      e.currentTarget.blur();
    }
    props.onFocus?.(e);
  };

  return (
    <PopoverPrimitive.Trigger
      data-slot="popover-trigger"
      data-testid="popover-trigger"
      data-as-child={asChild ? 'true' : 'false'}
      asChild={asChild}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      onFocus={handleFocus}
      {...props}
    />
  );
}

function PopoverContent({
  className,
  align = 'center',
  sideOffset = 0,
  onMouseEnter: propOnMouseEnter,
  onMouseLeave: propOnMouseLeave,
  ...props
}: React.ComponentProps<typeof PopoverPrimitive.Content>) {
  const { onMouseEnter: contextOnMouseEnter, onMouseLeave: contextOnMouseLeave } = useContext(PopoverContext);

  const handleMouseEnter = (e: React.MouseEvent<HTMLDivElement>) => {
    contextOnMouseEnter?.();
    propOnMouseEnter?.(e);
  };

  const handleMouseLeave = (e: React.MouseEvent<HTMLDivElement>) => {
    contextOnMouseLeave?.();
    propOnMouseLeave?.(e);
  };

  return (
    <PopoverPrimitive.Portal>
      <PopoverPrimitive.Content
        data-slot="popover-content"
        data-testid="popover-content"
        align={align}
        sideOffset={sideOffset}
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
        className={Libs.cn(
          'bg-popover text-popover-foreground data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 z-50 w-72 origin-(--radix-popover-content-transform-origin) rounded-md border p-4 shadow-[0px_4px_6px_0px_rgba(5,5,10,0.25)] shadow-[0px_10px_15px_0px_rgba(5,5,10,0.50)] outline-hidden mx-8 sm:mx-0',
          className,
        )}
        {...props}
      />
    </PopoverPrimitive.Portal>
  );
}

function PopoverAnchor({ ...props }: React.ComponentProps<typeof PopoverPrimitive.Anchor>) {
  return <PopoverPrimitive.Anchor data-slot="popover-anchor" {...props} />;
}

export { Popover, PopoverTrigger, PopoverContent, PopoverAnchor };
