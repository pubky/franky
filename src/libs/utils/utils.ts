import { clsx, type ClassValue } from 'clsx';
import { twMerge } from 'tailwind-merge';

type ExtractInitialsProps = { name: string; maxLength: number };
type CopyToClipboardProps = { text: string };
type FormatPublicKeyProps = { key: string; length: number };

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function formatPublicKey({ key, length = 12 }: FormatPublicKeyProps) {
  if (!key) return '';
  if (key.length <= length) return key;
  const prefix = key.slice(0, length / 2);
  const suffix = key.slice(-length / 2);
  return `${prefix}...${suffix}`;
}

export async function copyToClipboard({ text }: CopyToClipboardProps) {
  if (!navigator.clipboard) {
    throw new Error('Clipboard API not supported');
  }

  await navigator.clipboard.writeText(text);
}

// Helper function to normalize Radix UI IDs in container HTML for snapshot tests
export const normaliseRadixIds = (container: HTMLElement) => {
  const clonedContainer = container.cloneNode(true) as HTMLElement;
  const elementsWithAriaControls = clonedContainer.querySelectorAll('[aria-controls]');
  elementsWithAriaControls.forEach((el) => {
    const ariaControls = el.getAttribute('aria-controls');
    if (ariaControls && ariaControls.includes('radix-«r')) {
      el.setAttribute('aria-controls', 'radix-«r0»');
    }
  });
  return clonedContainer;
};

const customCases = [
  { name: 'bitcoin', color: '#FF9900' },
  { name: 'synonym', color: '#FF6600' },
  { name: 'bitkit', color: '#FF4400' },
  { name: 'pubky', color: '#C8FF00' },
  { name: 'blocktank', color: '#FFAE00' },
  { name: 'tether', color: '#26A17B' },
];

/**
 * Generates a consistent color for a given string
 * @param str - The string to generate a color for
 * @returns Hex color string
 */
export function generateRandomColor(str: string): string {
  const lowerStr = str.toLowerCase();

  // Check for cases
  for (const special of customCases) {
    if (lowerStr === special.name) {
      return special.color;
    }
  }

  // Generate a hash value from the input string
  const hash = Array.from(str).reduce((hash, char) => {
    return char.charCodeAt(0) + ((hash << 5) - hash);
  }, 0);

  // Ensure the hash is non-negative
  const positiveHash = Math.abs(hash);

  // Convert hash to a 2-character hex value
  const randomByte = positiveHash & 0xff; // extract the lowest 8 bits
  const randomHex = randomByte.toString(16).padStart(2, '0');

  // Pick a random pattern
  const patterns = [
    `FF00${randomHex}`,
    `FF${randomHex}00`,
    `${randomHex}FF00`,
    `${randomHex}00FF`,
    `00${randomHex}FF`,
    `00FF${randomHex}`,
  ];

  // Select pattern based on the hash
  const pattern = patterns[positiveHash % patterns.length];

  return `#${pattern}`;
}

/**
 * Converts a hex color to rgba format
 * @param hex - Hex color string (e.g., '#FF9900')
 * @param alpha - Alpha value (0-1)
 * @returns RGBA color string
 */
export function hexToRgba(hex: string, alpha: number) {
  const [r, g, b] = hex.match(/\w\w/g)!.map((x) => parseInt(x, 16));
  return `rgba(${r}, ${g}, ${b}, ${alpha})`;
}

export function extractInitials({ name, maxLength = 2 }: ExtractInitialsProps) {
  if (!name || typeof name !== 'string') return '';

  return name
    .trim()
    .split(/\s+/)
    .filter((word) => word.length > 0)
    .map((word) => word.charAt(0))
    .join('')
    .toUpperCase()
    .slice(0, maxLength);
}

export function formatInviteCode(code: string) {
  if (!code) return '';

  // Remove all non-alphanumeric characters
  const cleaned = code.replace(/[^a-zA-Z0-9]/g, '');

  // Convert to uppercase
  const uppercased = cleaned.toUpperCase();

  // If no valid characters, return empty string
  if (uppercased.length === 0) return '';

  // Handle different lengths
  if (uppercased.length <= 4) {
    return uppercased;
  } else if (uppercased.length <= 8) {
    return `${uppercased.slice(0, 4)}-${uppercased.slice(4)}`;
  } else {
    // Take only first 12 characters for longer codes
    const truncated = uppercased.slice(0, 12);
    return `${truncated.slice(0, 4)}-${truncated.slice(4, 8)}-${truncated.slice(8, 12)}`;
  }
}

export function clearCookies() {
  if (typeof document !== 'undefined') {
    document.cookie.split(';').forEach((cookie) => {
      const eqPos = cookie.indexOf('=');
      const name = eqPos > -1 ? cookie.substr(0, eqPos) : cookie;
      document.cookie = `${name.trim()}=;expires=Thu, 01 Jan 1970 00:00:00 GMT;path=/`;
    });
  }
}
