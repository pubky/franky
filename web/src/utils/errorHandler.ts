type ErrorType = 'NOT_FOUND' | 'VALIDATION' | 'DATABASE' | 'API' | 'NETWORK' | 'UNKNOWN';

interface ErrorDetails {
  type: ErrorType;
  message: string;
  context?: Record<string, unknown>;
  statusCode?: number;
}

export class ErrorHandler {
  static handle(error: unknown, details: ErrorDetails): never {
    // Log the error with context
    console.error(`[${details.type}] ${details.message}`, {
      error,
      context: details.context,
      statusCode: details.statusCode
    });

    // You can add different behaviors based on error type
    switch (details.type) {
      case 'NOT_FOUND':
        throw new Error(`Resource not found: ${details.message}`);
      
      case 'VALIDATION':
        throw new Error(`Validation error: ${details.message}`);
      
      case 'DATABASE':
        // You might want to do something specific for database errors
        // Like retrying the operation or cleaning up
        throw new Error(`Database error: ${details.message}`);

      case 'API':
        throw new Error(`API error (${details.statusCode}): ${details.message}`);

      case 'NETWORK':
        throw new Error(`Network error: ${details.message}`);
      
      default:
        throw new Error(`Unexpected error: ${details.message}`);
    }
  }

  static async handleFetch(
    url: string,
    options?: RequestInit,
    context?: Record<string, unknown>
  ): Promise<Response> {
    try {
      const response = await fetch(url, options);
      
      if (!response.ok) {
        return this.handle(null, {
          type: 'API',
          message: `Request failed for ${url}`,
          statusCode: response.status,
          context: {
            ...context,
            url,
            status: response.status,
            statusText: response.statusText
          }
        });
      }

      return response;
    } catch (error) {
      return this.handle(error, {
        type: 'NETWORK',
        message: `Network request failed for ${url}`,
        context: {
          ...context,
          url,
          error
        }
      });
    }
  }

  static handleAsync<T>(
    operation: () => Promise<T>,
    errorDetails: Omit<ErrorDetails, 'message'> & { message: string | ((error: unknown) => string) }
  ): Promise<T> {
    return operation().catch((error) => {
      const message = typeof errorDetails.message === 'function' 
        ? errorDetails.message(error)
        : errorDetails.message;

      return this.handle(error, { ...errorDetails, message });
    });
  }
} 