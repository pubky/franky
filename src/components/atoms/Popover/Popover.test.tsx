import { describe, it, expect } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { Popover, PopoverTrigger, PopoverContent } from './Popover';
import { normaliseRadixIds } from '@/libs/utils/utils';

describe('Popover', () => {
  it('renders popover trigger and content', () => {
    render(
      <Popover>
        <PopoverTrigger asChild>
          <button>Open Popover</button>
        </PopoverTrigger>
        <PopoverContent>
          <div>Popover Content</div>
        </PopoverContent>
      </Popover>,
    );

    const trigger = screen.getByRole('button', { name: /open popover/i });
    expect(trigger).toBeInTheDocument();
  });

  it('shows content when trigger is clicked', async () => {
    render(
      <Popover>
        <PopoverTrigger asChild>
          <button>Open Popover</button>
        </PopoverTrigger>
        <PopoverContent>
          <div>Popover Content</div>
        </PopoverContent>
      </Popover>,
    );

    const trigger = screen.getByRole('button');
    fireEvent.click(trigger);

    // Content should appear after click
    expect(screen.getByText('Popover Content')).toBeInTheDocument();
  });

  it('applies custom className to content', () => {
    render(
      <Popover>
        <PopoverTrigger asChild>
          <button>Open Popover</button>
        </PopoverTrigger>
        <PopoverContent className="custom-popover-class">
          <div>Popover Content</div>
        </PopoverContent>
      </Popover>,
    );

    const trigger = screen.getByRole('button');
    fireEvent.click(trigger);

    const content = screen.getByText('Popover Content').parentElement;
    expect(content).toHaveClass('custom-popover-class');
  });

  it('has proper data attributes', () => {
    render(
      <Popover>
        <PopoverTrigger asChild>
          <button>Open Popover</button>
        </PopoverTrigger>
        <PopoverContent>
          <div>Popover Content</div>
        </PopoverContent>
      </Popover>,
    );

    const trigger = screen.getByRole('button');
    expect(trigger).toHaveAttribute('data-slot', 'popover-trigger');

    fireEvent.click(trigger);

    const content = screen.getByText('Popover Content').parentElement;
    expect(content).toHaveAttribute('data-slot', 'popover-content');
  });
});

// Note: Radix UI generates incremental IDs (radix-«r0», radix-«r1», etc.) for aria-controls attributes.
// These IDs are deterministic within an identical test suite run but may change when a subset of tests are run or are run in a different order.
// Use normaliseRadixIds to ensure the snapshots are consistent.
describe('Popover - Snapshots', () => {
  it('matches snapshot for PopoverTrigger with default props', () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger>
          <button>Open Popover</button>
        </PopoverTrigger>
      </Popover>,
    );
    const normalizedContainer = normaliseRadixIds(container);
    expect(normalizedContainer.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for PopoverTrigger with asChild', () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger asChild>
          <button>Open Popover</button>
        </PopoverTrigger>
      </Popover>,
    );
    const normalizedContainer = normaliseRadixIds(container);
    expect(normalizedContainer.firstChild).toMatchSnapshot();
  });

  // TODO: address null output
  // it('matches snapshot for PopoverContent with default props', () => {
  //   const { container } = render(
  //     <Popover>
  //       <PopoverContent>
  //         <div>Popover Content</div>
  //       </PopoverContent>
  //     </Popover>,
  //   );
  //   const normalizedContainer = normalizeRadixIds(container);
  //   expect(normalizedContainer).toMatchSnapshot();
  // });

  // TODO: address null output
  // it('matches snapshots for PopoverContent with different configurations', () => {
  //   const { container: defaultContainer } = render(
  //     <Popover>
  //       <PopoverContent>
  //         <div>Default content</div>
  //       </PopoverContent>
  //     </Popover>,
  //   );
  //   expect(defaultContainer.firstChild).toMatchSnapshot();

  //   const { container: customClassContainer } = render(
  //     <Popover>
  //       <PopoverContent className="custom-popover-class">
  //         <div>Custom content</div>
  //       </PopoverContent>
  //     </Popover>,
  //   );
  //   expect(customClassContainer.firstChild).toMatchSnapshot();

  //   const { container: customAlignContainer } = render(
  //     <Popover>
  //       <PopoverContent align="start">
  //         <div>Start aligned content</div>
  //       </PopoverContent>
  //     </Popover>,
  //   );
  //   expect(customAlignContainer.firstChild).toMatchSnapshot();
  // });

  // TODO: address null output
  // it('matches snapshots for different content types', () => {
  //   const { container: simpleContainer } = render(
  //     <Popover>
  //       <PopoverContent>
  //         <div>Simple content</div>
  //       </PopoverContent>
  //     </Popover>,
  //   );
  //   expect(simpleContainer.firstChild).toMatchSnapshot();

  //   const { container: complexContainer } = render(
  //     <Popover>
  //       <PopoverContent>
  //         <div>
  //           <h3>Complex Content</h3>
  //           <p>With multiple elements</p>
  //           <button>Action</button>
  //         </div>
  //       </PopoverContent>
  //     </Popover>,
  //   );
  //   expect(complexContainer.firstChild).toMatchSnapshot();
  // });

  it('matches snapshot for complete popover structure', () => {
    const { container } = render(
      <Popover>
        <PopoverTrigger asChild>
          <button>Open Popover</button>
        </PopoverTrigger>
        <PopoverContent>
          <div>Popover Content</div>
        </PopoverContent>
      </Popover>,
    );
    const normalizedContainer = normaliseRadixIds(container);
    expect(normalizedContainer.firstChild).toMatchSnapshot();
  });
});
