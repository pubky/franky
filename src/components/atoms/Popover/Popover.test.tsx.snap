// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Popover - Snapshots > matches snapshot for PopoverTrigger with asChild 1`] = `
<button
  aria-controls="radix-«r0»"
  aria-expanded="false"
  aria-haspopup="dialog"
  data-as-child="true"
  data-slot="popover-trigger"
  data-state="closed"
  data-testid="popover-trigger"
  type="button"
>
  Open Popover
</button>
`;

exports[`Popover - Snapshots > matches snapshot for PopoverTrigger with default props 1`] = `
<button
  aria-controls="radix-«r0»"
  aria-expanded="false"
  aria-haspopup="dialog"
  data-as-child="false"
  data-slot="popover-trigger"
  data-state="closed"
  data-testid="popover-trigger"
  type="button"
>
  Open Popover
</button>
`;

exports[`Popover - Snapshots > matches snapshot for popover trigger in open state 1`] = `
<button
  aria-controls="radix-«r0»"
  aria-expanded="false"
  aria-haspopup="dialog"
  data-as-child="true"
  data-slot="popover-trigger"
  data-state="closed"
  data-testid="popover-trigger"
  type="button"
>
  Open Popover
</button>
`;
